pkg/il/interpreter/extern.go

```diff
@@ -0,0 +1,225 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package interpreter
+
+import (
+	"reflect"
+
+	"istio.io/mixer/pkg/il"
+)
+
+// Extern represents an external, native function that is callable from within the interpreter,
+// during program execution.
+type Extern struct {
+	name   string
+	pTypes []il.Type
+	rType  il.Type
+
+	v reflect.Value
+}
+
+// ExternFromFn creates a new, reflection based Extern, based on the given function. It panics if the
+// function signature is incompatible to be an extern.
+//
+// A function can be extern under the following conditions:
+// - Input parameter types are one of the supported types: string, bool, int64, float64, map[string]string.
+// - The return types can be:
+//   - none                                 (i.e. func (...) {...})
+//   - a supported type                     (i.e. func (...) string {...})
+//   - error                                (i.e. func (...) error {...})
+//   - suported type and error              (i.e. func (...) string, error {...})
+//
+func ExternFromFn(name string, fn interface{}) Extern {
+	t := reflect.TypeOf(fn)
+	if t.Kind() != reflect.Func {
+		panic("interpreter.ExternFromFn: not a function")
```

Comment:
I don't mind going either way but we should probably discuss first. I've been going back and forth on the error v.s. panic, and finally I've tried to settle on a model that Go seems to be following as well: https://github.com/golang/go/blob/master/src/reflect/type.go
Notice there are many methods here that simply panics. I believe the rationale is, most of these methods get called in a context where the caller either inherently knows that this is the right method to call (i.e. if you're calling reflect.Type.In(), you better make sure it is a function first), and/or have facilities to ensure before making the call (i.e. reflect.Type.Kind()).

Using a similar tact, I've tried to reduce the code paths to use panic where panicking indicates programming error (on our part). For this particular instance, my assumption is things like ExternFromFn (or its non-reflection counterpart) will be mostly called either by us, or by generated code. Trying to bubble these up will complicate the code.

Having said that, I don't mind making the change, I just want to make sure we establish the principle first.